# Name:Ritika lund
# CMS-ID:023-24-0262
# Section:H
# Github:https://github.com/Ritika-Lund/Machine-learning-Lab.git
# Kaggle:https://www.kaggle.com/ritikalund
# Dataset:Telco churn dataset
# Dataset source:https://www.kaggle.com/datasets/vedatgul/telco-churn-dataset

# Part 1:  Decision Tree Baseline Review 

# Task 1

In [4]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split

# Load cleaned dataset
df = pd.read_csv("clean_churn.csv")

print("Original shape:", df.shape)

# Check data
display(df.head())
print(df.columns)
df = df.drop_duplicates().reset_index(drop=True)

print("Shape after removing duplicates:", df.shape)
# Separate target
y = df["Churn"].map({"No": 0, "Yes": 1})

# Separate features
X = df.drop(columns=["Churn"])

print("X shape before encoding:", X.shape)
print("y shape:", y.shape)
# Find categorical columns
categorical_cols = X.select_dtypes(include=["object", "str"]).columns

# One-hot encoding
X = pd.get_dummies(
    X,
    columns=categorical_cols,
    dtype=int
)

print("X shape after encoding:", X.shape)
print("Missing values:", X.isnull().sum().sum())
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42,
    stratify=y
)

print("X_train:", X_train.shape)
print("X_test:", X_test.shape)
print("y_train:", y_train.shape)
print("y_test:", y_test.shape)

Original shape: (7043, 20)


,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,Female,0,Yes,No,1,No,No phone service,DSL,No,Yes,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,Male,0,No,No,34,Yes,No,DSL,Yes,No,Yes,No,No,No,One year,No,Mailed check,56.95,1889.50,No
2,Male,0,No,No,2,Yes,No,DSL,Yes,Yes,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,Male,0,No,No,45,No,No phone service,DSL,Yes,No,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,Female,0,No,No,2,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


Index(['gender', 'SeniorCitizen', 'Partner', 'Dependents', 'tenure',
       'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity',
       'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV',
       'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod',
       'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='str')
Shape after removing duplicates: (7021, 20)
X shape before encoding: (7021, 19)
y shape: (7021,)
X shape after encoding: (7021, 45)
Missing values: 0
X_train: (5616, 45)
X_test: (1405, 45)
y_train: (5616,)
y_test: (1405,)


# task 1 interpretation
The cleaned Telco Churn dataset was loaded and preprocessed using the same procedure as Lab 3. Categorical variables were converted into numerical features using one-hot encoding, resulting in 45 features. The dataset was divided into 80% training and 20% testing data using random_state=42 and stratified sampling to maintain the original class distribution.

# Task 2

In [5]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score
)


dt_model = DecisionTreeClassifier(
    max_depth=5,
    random_state=42
)

dt_model.fit(X_train, y_train)


dt_pred = dt_model.predict(X_test)

# Metrics
dt_accuracy = accuracy_score(y_test, dt_pred)
dt_precision = precision_score(y_test, dt_pred)
dt_recall = recall_score(y_test, dt_pred)
dt_f1 = f1_score(y_test, dt_pred)

print("Decision Tree Results")
print("Accuracy :", dt_accuracy)
print("Precision:", dt_precision)
print("Recall   :", dt_recall)
print("F1-score :", dt_f1)

Decision Tree Results
Accuracy : 0.7943060498220641
Precision: 0.6202898550724638
Recall   : 0.5752688172043011
F1-score : 0.596931659693166


# task 2 interpretation
The Decision Tree achieved approximately 79.4% accuracy and an F1-score of 59.7%. Although its overall accuracy is reasonable, its recall indicates that the model misses a considerable portion of customers who actually churn. Therefore, additional models should be evaluated to determine whether churn prediction can be improved.

# Task 3

In [6]:
from sklearn.ensemble import RandomForestClassifier


rf_model = RandomForestClassifier(
    n_estimators=200,
    random_state=42,
    n_jobs=-1
)

rf_model.fit(X_train, y_train)

rf_pred = rf_model.predict(X_test)

print("Random Forest trained successfully.")

Random Forest trained successfully.


# Task 3 interpretation
A Random Forest was trained using 200 decision trees. The model combines predictions from multiple trees rather than relying on a single tree, which can help reduce overfitting and improve generalization to unseen data.

# Task 4
Random Forest combines multiple decision trees and averages their predictions, which reduces the variance of an individual tree. Because the trees are trained with different samples and feature subsets, the ensemble is generally less sensitive to the overfitting of any single tree and can therefore generalize better to unseen data.